### ISSUE TITLE
[Suggestions] [1.18.2+] - Biome search and server-wide map

### ISSUE THREAD

**Issue by Eleiyas** [TeamJM/journeymap#419]
MC 1.18.2+

1) Biome search bar. 
We can already search out specific coordinates, but it would also be nice to be able to search out a biome name, to see if we've already visited one. This is good for the MC Advancement and Biomes O' Plenty. It's also helpful on servers that are using any biome-adding mod(s).

Pros: good for biome-adding mods and advancement hunters
Cons: none?

2) Server-wide map
I'd like to see, that when JourneyMap is installed on a server, there can be an option to have everyone's maps pool together into one main server map. 

Example: 
everyone has their bases in a different corner of the map, say 10,000/-10,000 // -10,000/10,000 // 10,000/10,000 // -10,000/-10,000
instead of all players needing to explore those hundreds of thousands of blocks, for them to all show up on their Map, it would pool all players into one main map, requiring far less (needless) exploration. 

Pros: would be great for servers that are focussed on a lot of exploration and/or have ginormous maps
Cons:
The main issues I see are that because JM saves and loads maps locally from the player's PC (as it's mainly client-side), it would require a re-write so that the map gets saved to the server destination, then downloaded by each player, likely requiring some beefy netcode implementation. Not to mention that for very large servers with (let's face it) more than ~25 players and huge maps, this would likely cause server slow-down and wouldn't be great for players with bad internet.

**Comment by mysticdrew** [comment]
1. Will never happen. "search specific coordinates" is not really a search. It is a calculation. 
2. Maybe.

**Comment by Eleiyas** [comment]
> 1. Will never happen. "search specific coordinates" is not really a search. It is a calculation. You can see the biomes as you explore. We have no way to get the biome for chunks that are not generated.
> 2. Maybe.

Ahh, I think you might have misunderstood me for my first idea. When you explore somewhere and it's logged in the map, you can see the coordinates and the biome when you hover over it.

What I would like is a bar that can search for a biome that I've already visited (to double check if I have or not). For non-explored areas it could return a null / error of "region unexplored" or something.

**Comment by mysticdrew** [comment]
> > 1. Will never happen. "search specific coordinates" is not really a search. It is a calculation. You can see the biomes as you explore. We have no way to get the biome for chunks that are not generated.
> > 2. Maybe.
> 
> Ahh, I think you might have misunderstood me for my first idea. When you explore somewhere and it's logged in the map, you can see the coordinates and the biome when you hover over it.
> 
> What I would like is a bar that can search for a biome that I've already visited (to double check if I have or not). For non-explored areas it could return a null / error of "region unexplored" or something.

No... that is not really possible with how the data is stored. All that is doing, is reading from the in memory chunk. 
Imagine having the query over every single block in the game... A chunk is 16x16 blocks, a region is 32x32 chunks. For one region that is 262,144 blocks (since we'd only need to check the current y level). Unless you want to add in cave biomes into that search then that is 100,663,296 blocks. 

If you or someone can come up with a rapid algorithm to do it, I can then look into implementing it. But otherwise, I will not even entertain the idea. Sorry.

**Comment by Eleiyas** [comment]
> If you or someone can come up with a rapid algorithm to do it, I can then look into implementing it. But otherwise, I will not even entertain the idea. Sorry.

I find it interesting that it does work like that, because I can boot up my server, load into it, open up JourneyMap and scroll across my explored areas and it shows coords and biome, which would tell me that that particular information is stored somewhere, no?

I mean, how does the "Nature's Compass" mod do it then, because that can also search unexplored areas of the map for specific biomes?

Either way, I'm just genuinely curious and if you say it's not feasible, then it's not feasible :) I do find it funny that one suggestion that I thought would be easy is the difficult one and the one I thought would be hard you gave a "maybe" to :P

**Comment by mysticdrew** [comment]
I have no idea how they do it, I have never heard of that mod. I suppose you could suggest that they add journeymap integration via our extensive API. It it could mark biomes on the map.

**Comment by mysticdrew** [comment]
> I find it interesting that it does work like that, because I can boot up my server, load into it, open up JourneyMap and scroll across my explored areas and it shows coords and biome, which would tell me that that particular information is stored somewhere, no?

Yes in our block data cache, but it is not stored in a searchable manner. When the mouse is over a pixel on the map, we have a calculation to determine the blockpos, and we do a simple lookup for the biome for that block.

**Comment by Eleiyas** [comment]
> > I find it interesting that it does work like that, because I can boot up my server, load into it, open up JourneyMap and scroll across my explored areas and it shows coords and biome, which would tell me that that particular information is stored somewhere, no?
> 
> Yes in our block data cache, but it is not stored in a searchable manner. When the mouse is over a pixel on the map, we have a calculation to determine the blockpos, and we do a simple lookup for the biome for that block.

and it's impossible for you to save it in a searchable manner? I guess that's what you're saying, right?

**Comment by mysticdrew** [comment]
not impossible but currently it is not. I have been thinking about it a bit. 
Let's see if that natures compass adds integration. I think that could be a good thing. 
I wont be closing this ticket as I do I like the idea of it, but it is not a priority currently.

**Comment by mysticdrew** [comment]
So, good news. 
I will be added some type of biome locator to either journeymap or [journeymap-additions](https://github.com/TeamJM/journeymap-additions) (I havent decided where it fits best yet). Minecraft has a /locatebiome command that I did not know about. This will make it much easier for me to implement.
